### PR TITLE
Table: custom column header components

### PR DESCRIPTION
**Pull request by gtk-grafana**
**What is this feature?**
Specific usages of the table might have additional functionality that need to be exposed to end users. This PR provides the ability for developers to pass in a map of react elements that will (hopefully) facilitate more ways for users to visualize and interact with their data!

**Why do we need this feature?**
So we can add interactive menus for logs and other table implementations. 

**Who is this feature for?**
Developers building interactive UIs with table.

**Special notes for your reviewer:**
Example usage:
```
  const map = new Map();
  map.set(0, <Icon name={'ellipsis-v'} />)

  return (
    <Table
      columnHeaders={map}
      data={tableFrame}
     // {...tableProps}
    />
  );
```
Which will set the icon relatively positioned after any other table header icons.
image

However if you want to absolutely position within the cell, that's ok too, as long as you leave some space for the resize-handlers (if active).
image

Although the absolute positioned solution might want additional right padding on the table cell to keep content from overlapping with menu, but that can be addressed in a later PR (maybe we want to right-align icons instead of absolutely positioning things, etc)

Opening this PR to start the conversation, I'm not wed to any particular implementation and would love to hear any other ideas!

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/Table/types.ts [hunk 1]
```diff
@@ -99,6 +99,8 @@ export interface Props {
   enableSharedCrosshair?: boolean;
   // The index of the field value that the table will initialize scrolled to
   initialRowIndex?: number;
+  // Map of react elements to render within column headers, index is field id
+  columnHeaders?: Map<number, React.ReactElement>;
```
Comment (torkelo) [inline packages/grafana-ui/src/components/Table/types.ts:103]: why use different pattern for passing this than we do for cellComponent? 

https://github.com/grafana/grafana/blob/cfcf03bf7af4f1928ed728aa38eb86c874ba46cc/packages/grafana-ui/src/components/Table/Table.story.tsx#L286

**Review by torkelo — CHANGES_REQUESTED** [review]
I think this should follow the pattern established by cellComponent, unless there is a good reason not to?

File: packages/grafana-ui/src/components/Table/HeaderRow.tsx [hunk 2]
```diff
@@ -31,17 +32,25 @@ export const HeaderRow = (props: HeaderRowProps) => {
...
           >
-            {headerGroup.headers.map((column: Column, index: number) =>
-              renderHeaderCell(column, tableStyles, showTypeIcons)
-            )}
+            {headerGroup.headers.map((column: Column, index: number) => {
+              if (columnHeaders && columnHeaders.has(index)) {
+                return renderHeaderCell(column, tableStyles, showTypeIcons, columnHeaders.get(index));
+              }
+              return renderHeaderCell(column, tableStyles, showTypeIcons);
+            })}
```
Comment (matyax) [inline packages/grafana-ui/src/components/Table/HeaderRow.tsx:40]: You could skip the check and renderHeaderCell() would get `undefined` or the column header. 
```suggestion
            {headerGroup.headers.map((column: Column, index: number) => (
              renderHeaderCell(column, tableStyles, showTypeIcons, columnHeaders.get(index));
            ))}
```

**Comment by gtk-grafana** [comment]
> I think this should follow the pattern established by cellComponent, unless there is a good reason not to?

@torkelo I'm having a hard time figuring out the cellComponent, does it allow us to add additional components to the column headers? From what I can see it would only allow us to define custom fields and allow us to customize the cells, not the headers?

**Comment by torkelo** [comment]
@gtk-grafana I said it should follow the same pattern (same technical solution) as that of cellComponent. We of course would need a new field config option named "headerComponent"  

the cellComponent fieldConfig option allows any user of Table component or scenes TablePanel to define a custom react component that will handle rendering the cell. Since this is a field config option it can be defined per field (column)

**Comment by torkelo** [comment]
@gtk-grafana but I understand it's not trivial to grock all this and how to leverage the data driven config system (Field config).  

Built a small POC here to explain what I mean , https://github.com/grafana/grafana/pull/83254

**Comment by gtk-grafana** [comment]
https://github.com/grafana/grafana/pull/83254 appears to work fine as a solution and is a better implementation.

**Comment by gtk-grafana** [comment]
Thank you @torkelo! Yeah let's pursue the solution in your PR, closing!

**Pull request closed without merging**